### PR TITLE
spotify-unwrapped: add macOS builds

### PR DESCRIPTION
**Pull request by Enzime**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -31899,7 +31899,10 @@ with pkgs;
 
   spotify-unwrapped = callPackage ../applications/audio/spotify { };
 
-  spotify = callPackage ../applications/audio/spotify/wrapper.nix { };
+  spotify = if stdenv.isDarwin then
+    spotify-unwrapped
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:31903]: This should be done one level higher.

File: pkgs/applications/audio/spotify/linux.nix [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  # spotify ourselves:
+  # https://community.spotify.com/t5/Desktop-Linux/Redistribute-Spotify-on-Linux-Distributions/td-p/1695334
+  src = fetchurl {
+    url = "https://api.snapcraft.io/api/v1/snaps/download/pOBIoZ2LrCB3rDohMxoYGnbN14EHOgD7_${rev}.snap";
+    sha512 = "1209b956822d8bb661daa2c88616bed403ec26dc22c6b866cecff59235c56112284c2f99aa06352fc0df6fcd15225a6ad60afd3b4ff4d7b948ab83e70ab31a71";
+  };
+
+  nativeBuildInputs = [ makeWrapper wrapGAppsHook squashfsTools ];
+
+  dontStrip = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/spotify/linux.nix:88]: Why do we need this?

File: pkgs/applications/audio/spotify/linux.nix [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+      mv ./usr/* $out/
+
+      cp meta/snap.yaml $out
+
+      # Work around Spotify referring to a specific minor version of
+      # OpenSSL.
+
+      ln -s ${lib.getLib openssl}/lib/libssl.so $libdir/libssl.so.1.0.0
+      ln -s ${lib.getLib openssl}/lib/libcrypto.so $libdir/libcrypto.so.1.0.0
+      ln -s ${nspr.out}/lib/libnspr4.so $libdir/libnspr4.so
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/spotify/linux.nix:131]: All those should use lib.getLib

File: pkgs/applications/audio/spotify/linux.nix [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        ''${gappsWrapperArgs[@]} \
+        --prefix LD_LIBRARY_PATH : "$librarypath" \
+        --prefix PATH : "${gnome.zenity}/bin"
+
+      # fix Icon line in the desktop file (#48062)
+      sed -i "s:^Icon=.*:Icon=spotify-client:" "$out/share/spotify/spotify.desktop"
+
+      # Desktop file
+      mkdir -p "$out/share/applications/"
+      cp "$out/share/spotify/spotify.desktop" "$out/share/applications/"
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/spotify/linux.nix:154]: Why not symlink?

File: pkgs/applications/audio/spotify/darwin.nix [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+}:
+
+stdenv.mkDerivation {
+  inherit pname meta;
+
+  version = "1.1.97.962.g24733a46";
+
+  src = if stdenv.isAarch64 then (
+    fetchurl {
+      url = "https://download.scdn.co/SpotifyARM64.dmg";
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/spotify/darwin.nix:15]: This must go through web.archive.org

**Comment by Enzime** [comment]
@SuperSandro2000 all I did as part of this PR was move the contents from `default.nix` to `linux.nix`, what do you think about fixing these issues in a different PR to keep this one focused?

**Comment by laMudri** [comment]
I haven't used this package for years. Please could I be removed from the maintainers list?

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (Enzime) [reply pkgs/top-level/all-packages.nix:31903]: What level are you suggesting?

**Review by Atemu — APPROVED** [review]
Result of `nixpkgs-review pr 195473` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>spotify (spotify-unwrapped)</li>
  </ul>
</details>

Opened it and it works as expected. I'm on Monterey though, so idk whether it'd work with Ventura's new Gatekeeper but we're not modifying it, so it should.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:31903]: I mean one level lower.

It is a bit awkward to get a wrapped package on Linux and an unwrapped on darwin from the same attribute. One a second look we could remove that wrapper entirely and move the one functionality into the spotify package itself then we also don't need to make this awkward condition here.

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 all I did as part of this PR was move the contents from `default.nix` to `linux.nix`, what do you think about fixing these issues in a different PR to keep this one focused?

GitHub has no support for displaying moved hunks like the git cli so unless I compare the lines one by one I do not know that. Fine for me but the wrapper condition needs to be less awkward between Linux and Darwin. Let me create a PR that removes the wrapper.

**Comment by SuperSandro2000** [comment]
#199779

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (Atemu) [reply pkgs/top-level/all-packages.nix:31903]: It's not really unwrapped, it's an entirely different package. It just happens to be called the same way the unwrapped linux package is and it's simplest to make that attr an alias.

**Comment by SuperSandro2000** [comment]
I just merged the consolidation of the wrapper. Now we don't need to think about the differences between darwin and linux and can make the package decision in top-level.nix or in the package. I have a preference to do it inside of the package but the other way is also fine.

**Comment by Enzime** [comment]
@SuperSandro2000 rebased on top of master :+1:

File: pkgs/applications/audio/spotify/default.nix [hunk 6]
```diff
@@ -1,181 +1,18 @@
...
-    maintainers = with maintainers; [ eelco ftrvxmtrx sheenobu mudri timokau ma27 ];
-    platforms = [ "x86_64-linux" ];
+    maintainers = with maintainers; [ eelco ftrvxmtrx sheenobu timokau ma27 ];
+    platforms = [ "x86_64-linux" "x86_64-darwin" "aarch64-darwin" ];
   };
-}
+
+in if stdenv.isDarwin
+then callPackage ./darwin.nix { inherit pname meta; }
+else callPackage ./linux.nix { inherit pname meta; }
```
Comment (Atemu) [inline pkgs/applications/audio/spotify/default.nix:18]: Do this in all-packages.

Comment (Enzime) [reply pkgs/applications/audio/spotify/default.nix:18]: The majority of apps I've seen in nixpkgs do it this way:

https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/web/postman/default.nix#L17-L19
https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/jetbrains/default.nix#L23
https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/version-management/p4v/default.nix#L22-L23
https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/instant-messengers/discord/default.nix#L60-L63

File: pkgs/applications/audio/spotify/default.nix [hunk 7]
```diff
@@ -1,181 +1,18 @@
...
-
-      runHook postInstall
-    '';
 
   meta = with lib; {
     homepage = "https://www.spotify.com/";
     description = "Play music from the Spotify music service";
     sourceProvenance = with sourceTypes; [ binaryNativeCode ];
     license = licenses.unfree;
-    maintainers = with maintainers; [ eelco ftrvxmtrx sheenobu mudri timokau ma27 ];
```
Comment (Atemu) [inline pkgs/applications/audio/spotify/default.nix:178]: Why?

Comment (Enzime) [reply pkgs/applications/audio/spotify/default.nix:178]: mudri requested to be removed earlier in this PR

https://github.com/NixOS/nixpkgs/pull/195473#issuecomment-1299295473

Comment (Atemu) [reply pkgs/applications/audio/spotify/default.nix:178]: Could you add that to the commit message?

Comment (Enzime) [reply pkgs/applications/audio/spotify/default.nix:178]: Done :+1:

**Comment by yamashitax** [comment]
##### Reviewed points

- [x] package name fits guidelines
- [x] package version fits guidelines
- [ ] package build on aarch64-darwin
- [ ] executables tested on aarch64-darwin
- [ ] all depending packages build

##### Possible improvements

Unable to build due to a merge conflict.

##### Comments

Result of `nixpkgs-review`

```
➜ nix-shell -p nixpkgs-review --run "nixpkgs-review pr 195473"
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/195473/head:refs/nixpkgs-review/1
remote: Enumerating objects: 24, done.
remote: Counting objects: 100% (22/22), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 12 (delta 9), reused 6 (delta 5), pack-reused 0
Unpacking objects: 100% (12/12), 1.54 KiB | 142.00 KiB/s, done.
From https://github.com/NixOS/nixpkgs
   8f8bb3bb7ed..ed7e3a25a9d  master     -> refs/nixpkgs-review/0
$ git worktree add /Users/yamashita/.cache/nixpkgs-review/pr-195473-1/nixpkgs ed7e3a25a9d964177bd52ff31238bd950dca6807
Preparing worktree (detached HEAD ed7e3a25a9d)
Updating files: 100% (34527/34527), done.
HEAD is now at ed7e3a25a9d Merge pull request #226285 from marsam/update-flexget
$ git merge --no-commit --no-ff 5f78d39caa98d86f1b29caee5733ff4febe0141e
Auto-merging pkgs/applications/audio/spotify/default.nix
CONFLICT (content): Merge conflict in pkgs/applications/audio/spotify/default.nix
Auto-merging pkgs/applications/audio/spotify/update.sh
Automatic merge failed; fix conflicts and then commit the result.
https://github.com/NixOS/nixpkgs/pull/195473 failed to build
$ git worktree prune
```

**Comment by Enzime** [comment]
@yamashitax updated :+1:

File: pkgs/applications/audio/spotify/default.nix [hunk 8]
```diff
@@ -22,54 +22,6 @@ let
...
-    xorg.libxshmfence
-    xorg.libXtst
-    zlib
-  ];
-
-in
-
-stdenv.mkDerivation {
   pname = "spotify";
   inherit version;
```
Comment (yamashitax) [inline pkgs/applications/audio/spotify/default.nix:26]: Already defined on Line 16

**Review by yamashitax — CHANGES_REQUESTED** [review]

File: pkgs/applications/audio/spotify/default.nix (hunk 8 above)
Comment (Enzime) [reply pkgs/applications/audio/spotify/default.nix]: Good catch :+1: I messed up the rebase

**Comment by Enzime** [comment]
@yamashitax updated :+1: if you could review again that would be much appreciated

**Review by yamashitax — APPROVED** [review]
Result of `nixpkgs-review pr 195473` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>spotify</li>
  </ul>
</details>

Works as intended

**Comment by yamashitax** [comment]
@SuperSandro2000 Seems fine to merge this now, from what I can tell.

File: pkgs/applications/audio/spotify/default.nix [hunk 9]
```diff
@@ -1,182 +1,17 @@
...
-    '';
 
   meta = with lib; {
     homepage = "https://www.spotify.com/";
     description = "Play music from the Spotify music service";
     sourceProvenance = with sourceTypes; [ binaryNativeCode ];
     license = licenses.unfree;
-    maintainers = with maintainers; [ eelco ftrvxmtrx sheenobu mudri timokau ma27 ];
-    platforms = [ "x86_64-linux" ];
+    maintainers = with maintainers; [ eelco ftrvxmtrx sheenobu timokau ma27 ];
```
Comment (Atemu) [inline pkgs/applications/audio/spotify/default.nix:11]: Could you remove these people from the Darwin meta and add yourself instead? They only maintain the Linux package.

Comment (Enzime) [reply pkgs/applications/audio/spotify/default.nix]: Done :+1:

**Review by Atemu — APPROVED** [review]
Result of `nixpkgs-review pr 195473` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>spotify</li>
  </ul>
</details>

**Comment by Atemu** [comment]
(This is on Ventura; works with Gatekeeper.)

File: pkgs/applications/audio/spotify/darwin.nix [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+, pname
+, meta
+, fetchurl
+, undmg
+, lib
+}:
+
+stdenv.mkDerivation {
+  inherit pname;
+
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/spotify/darwin.nix:11]:
```suggestion
```

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by jtojnar** [comment]
> GitHub has no support for displaying moved hunks like the git cli so unless I compare the lines one by one I do not know that. Fine for me but the wrapper condition needs to be less awkward between Linux and Darwin. Let me create a PR that removes the wrapper.

That is why it is good idea to rename files in a separate commit from any other changes. That way the change commit can be reviewed separately and `git blame` will continue to work.

**Pull request merged**